### PR TITLE
DR-2504 OIDC config support

### PR DESCRIPTION
**Pull request by nmalfroy**
This PR does the following:
- Pulls in the new Apache proxy that we should be using now to allow talking to B2C
- Allows poking through the /oauth2 path to support auth hosted by TDR

### CODE REVIEW

File: charts/oidc-proxy/templates/configmap.yaml [hunk 1]
```diff
@@ -12,12 +12,14 @@ data:
     ServerTokens ProductOnly
     TraceEnable off
 
-    LogFormat "%h %l %u \"%{OIDC_CLAIM_email}i\" \"%{X-App-ID}i\" [%{%FT%T}t.%{msec_frac}t%{%z}t] %D %X \"%r\" %>s %b \"%{Referer}i\" \"%{User-Agent}i\"" combined
-    LogFormat "%{X-Forwarded-For}i %l %u \"%{OIDC_CLAIM_email}i\" \"%{X-App-ID}i\" [%{%FT%T}t.%{msec_frac}t%{%z}t] %D %X \"%r\" %>s %b \"%{Referer}i\" \"%{User-Agent}i\"" proxy
+    LogFormat "%h %l %u \"%{OAUTH2_CLAIM_email}i\" \"%{X-App-ID}i\" [%{%FT%T}t.%{msec_frac}t%{%z}t] %D %X \"%r\" %>s %b \"%{Referer}i\" \"%{User-Agent}i\"" combined
```
Comment (rtitle) [inline charts/oidc-proxy/templates/configmap.yaml:15]: Just fyi, AppSec wanted to add `\"%{Origin}i\"` to the log line. I added it between Referer and User-Agent in the other proxies. See: https://broadworkbench.atlassian.net/browse/DDO-1983, [Sam site.conf](https://github.com/broadinstitute/firecloud-develop/blob/dev/base-configs/sam/site.conf.ctmpl#L98).

Not sure if it's needed for TDR or how your log infrastructure works. :)

File: charts/oidc-proxy/templates/configmap.yaml [hunk 2]
```diff
@@ -95,7 +114,23 @@ data:
...
+                # An example expression can look like:
+                # Require expr "%{REMOTE_USER} != %{REMOTE_USER} || %{HTTP:OAUTH2_CLAIM_email} =~ /.gserviceaccount.com$/"
+                Require expr "%{REMOTE_USER} != %{REMOTE_USER} || ('${ENVIRONMENT}' in {'dev', 'alpha', 'perf', 'staging', 'prod'} && (${ALLOW_EXPRESSION}))"
+              </RequireAll>
+            </RequireAll>
+
+
+            RequestHeader set oidc_claim_email "expr=%{HTTP:OAUTH2_CLAIM_email}"
+            RequestHeader set oidc_claim_user_id "expr=%{HTTP:OAUTH2_CLAIM_sub}"
+            RequestHeader set oidc_access_token "expr=%{HTTP:OAUTH2_CLAIM_access_token}"
```
Comment (rtitle) [inline charts/oidc-proxy/templates/configmap.yaml:133]: Another nit: I added
```
RequestHeader set oidc_claim_expires_in "expr=%{HTTP:OAUTH2_CLAIM_exp}"
```
here -- some of the services used this header.

File: charts/oidc-proxy/templates/configmap.yaml [hunk 3]
```diff
@@ -95,7 +114,23 @@ data:
...
+            <RequireAll>
+              <RequireAll>
+                ${AUTH_REQUIRE2}
+              </RequireAll>
+              <RequireAll>
+                # Note: we need to reference %{REMOTE_USER} to ensure that this expression is evaluated _after_ the authentication has happened
+                # %{REMOTE_USER} != %{REMOTE_USER} will never evaluate to true but will ensure that the expression is evaluated at the correct time
+                # An example expression can look like:
+                # Require expr "%{REMOTE_USER} != %{REMOTE_USER} || %{HTTP:OAUTH2_CLAIM_email} =~ /.gserviceaccount.com$/"
+                Require expr "%{REMOTE_USER} != %{REMOTE_USER} || ('${ENVIRONMENT}' in {'dev', 'alpha', 'perf', 'staging', 'prod'} && (${ALLOW_EXPRESSION}))"
```
Comment (rtitle) [inline charts/oidc-proxy/templates/configmap.yaml:126]: Curious if you set `$ALLOW_EXPRESSION` anywhere?

I was having trouble with this in some of the terra-helmfile proxies. The old proxy used the jq-based `Require claims_expr` which isn't possible in mod_oauth2. In the firecloud-develop proxies we are building the `Require expr` with go ctmpl logic. But for the helmfile configmap we can't use ctmpl on a rendered k8s secret (as far as I can tell). So kind of stuck on how to build this expression…

File: charts/oidc-proxy/values.yaml [hunk 4]
```diff
@@ -4,9 +4,11 @@
 
 replicaCount: 1
 
+## Versions can be found here:
+## https://github.com/broadinstitute/openidc-terra-proxy/tags
 image:
-  repository: broadinstitute/openidc-proxy
-  version: bernick_tcell
+  repository: us.gcr.io/broad-dsp-gcr-public/openidc-terra-proxy
+  version: v0.1.12
```
Comment (rtitle) [inline charts/oidc-proxy/values.yaml:11]: Cool, this is the latest version 👍

File: charts/oidc-proxy/templates/configmap.yaml (hunk 1 above)
Comment (nmalfroy) [reply charts/oidc-proxy/templates/configmap.yaml:15]: Ooh good call.  I can totally add that

File: charts/oidc-proxy/templates/configmap.yaml (hunk 3 above)
Comment (nmalfroy) [reply charts/oidc-proxy/templates/configmap.yaml:126]: Honestly, we don't even really need it.  I had it there more of as an example but realistically we depend on Sam to kick out users that don't meet the required clientid

File: charts/oidc-proxy/values.yaml (hunk 4 above)
Comment (nmalfroy) [reply charts/oidc-proxy/values.yaml:11]: Yep!

File: charts/oidc-proxy/templates/configmap.yaml (hunk 3 above)
Comment (rtitle) [reply charts/oidc-proxy/templates/configmap.yaml:126]: That's true -- a lot of the checks in the proxy are redundant if the service uses Sam. This probably goes for the LDAP check too..

File: charts/oidc-proxy/values.yaml [hunk 5]
```diff
@@ -70,3 +72,6 @@ ingress:
     # - secretName: datarepo-example-tls
     #   hosts:
     #     - datarepo.example.com
+
+oidc:
+  b2cMetadataEndpoint: https://terradevb2c.b2clogin.com/terradevb2c.onmicrosoft.com/B2C_1A_SIGNUP_SIGNIN/v2.0/.well-known/openid-configuration
```
[the remark this answers is not in the document]
Comment (nmalfroy) [reply charts/oidc-proxy/values.yaml:77]: Yeah, I was torn. Right now we're just doing dev for all of the environments except prod so this seemed like the least painful way (at least for now)

[the remark this answers is not in the document]
Comment (samanehsan) [reply charts/oidc-proxy/values.yaml:77]: Got it! That makes sense to me in that case...less places to change 😅

**Pull request merged**